### ISSUE TITLE
engine_started called multi times in one process, if multi spiders exists.

### ISSUE THREAD

**Issue by ultragis** [scrapy/scrapy#2357]
My Extension code is:

```
class SpiderStatsExtension(object):
    @classmethod
    def from_crawler(cls, crawler):
        if not hasattr(cls, '__instance'):
            setattr(cls, '__instance', cls())
        ext = getattr(cls, '__instance')
        crawler.signals.connect(ext.engine_started, signal=signals.engine_started)
        crawler.signals.connect(ext.engine_stopped, signal=signals.engine_stopped)
        crawler.signals.connect(ext.item_scraped, signal=signals.item_scraped)
        crawler.signals.connect(ext.spider_error, signal=signals.spider_error)
        return ext
```

My process start code is:

```
process = CrawlerProcess(get_project_settings())
spiders = process.spider_loader.list()
for spider in spiders:
    process.crawl(process.spider_loader.load(spider))
process.start()
```

I found the from_crawler called multi times if my project have multi spiders.
Is it expectant design? 
I need once call for on process to record stats info, how can i do for it?
thanks.

**Comment by kmike** [comment]
Hi @ultragis,

Yes, this is an expected design. engine_started signal (as well as all other signals) are issued per-crawler, i.e. per-spider; CrawlerProcess is a relatively small helper to run these crawlers/spiders in parallel.

from_crawler method receives Crawler instance, not CrawlerProcess; there is 1-to-1 relationship between Crawler and Spider. Stats are also per-spider, not global CrawlerProcess. Spider and Crawler are not aware of CrawlerProcess.

**Comment by ultragis** [comment]
thanks. 
But I can't get process startup and stopped signal.
So, I add startup and stopped statement before and after process.start(), not perfect but can be used.

**Comment by Digenis** [comment]
Since multi spider crawls were discontinued,
I thought there's no need to allocate per-spider resources in extensions.
Extension instances are still associated with a single crawler, aren't they?

**Comment by ultragis** [comment]
Sorry, I don't understand your main, are you speak to me?
I just want a statistics for multi spiders in one Process.

**Comment by cathalgarvey** [comment]
Hi @ultragis - Have you found a solution to your problem here? This issue is pretty stale and already marked as `wontfix`, so I'll close it for now. But feel free to reopen if you feel there's a new direction to consider. Thank you!
